Index: mtd-utils-2.2.1/ubifs-utils/mkfs.ubifs/mkfs.ubifs.h
===================================================================
--- mtd-utils-2.2.1.orig/ubifs-utils/mkfs.ubifs/mkfs.ubifs.h
+++ mtd-utils-2.2.1/ubifs-utils/mkfs.ubifs/mkfs.ubifs.h
@@ -53,7 +53,7 @@
 #include <errno.h>
 #include <libgen.h>
 #include <ctype.h>
-#include <uuid.h>
+#include <uuid/uuid.h>
 #include <sys/file.h>
 
 #ifdef WITH_CRYPTO
